powertools-parameters/src/main/java/software/amazon/lambda/powertools/parameters/BaseProvider.java

```diff
@@ -0,0 +1,195 @@
+     * If you need a more complex transformation (to Object), use {@link #get(String, Class)} method instead of this one. <br/>
+     *
+     * @param key key of the parameter
+     * @return the String value of the parameter
+     */
+    public String get(String key) {
+        if (hasNotExpired(key)) {
+            resetToDefaults();
+            return (String) store.get(key);
+        }
+
+        String value = getValue(key);
+
+        String transformedValue = value;
+        if (transformerClass != null) {
+            if (!BasicTransformer.class.isAssignableFrom(transformerClass)) {
+                resetToDefaults();
+                throw new IllegalArgumentException("Wrong Transformer for a String, choose a BasicTransformer");
+            }
+            try {
+                BasicTransformer transformer = (BasicTransformer) transformerClass.newInstance();
+                transformedValue = transformer.applyTransformation(value);
+            } catch (InstantiationException | IllegalAccessException e) {
+                resetToDefaults();
+                throw new TransformationException(e);
+            }
+        }
+
+        store.put(key, transformedValue, Instant.now().plus(maxAge, ChronoUnit.SECONDS).toEpochMilli());
+
+        resetToDefaults();
+        return transformedValue;
+    }
+
+    /**
+     * Get the value of a parameter, either from the underlying store or a cached value (if not expired).<br/>
+     * Using this method, you must apply a transformation (eg. json/xml to Object). <br/>
+     * Set a {@link Transformer} with {@link #withTransformation(Class)}.<br/><br/>
+     * If you need a simpler transformation (to String), use {@link #get(String)} method instead of this one.
+     *
+     * @param key         key of the parameter
+     * @param targetClass class of the target Object (after transformation)
+     * @return the Object (T) value of the parameter
+     */
+    public <T> T get(String key, Class<T> targetClass) {
+        if (hasNotExpired(key)) {
+            resetToDefaults();
+            return (T) store.get(key);
+        }
+
+        String value = getValue(key);
+
+        if (transformerClass == null) {
+            resetToDefaults();
+            throw new IllegalArgumentException("transformer is null, use withTransformation to specify a transformer");
+        }
+        Transformer<T> transformer;
+        try {
+            transformer = transformerClass.newInstance();
+        } catch (InstantiationException | IllegalAccessException e) {
+            resetToDefaults();
+            throw new TransformationException(e);
+        }
+        T transformedValue = transformer.applyTransformation(value, targetClass);
+
+        store.put(key, transformedValue, Instant.now().plus(maxAge, ChronoUnit.SECONDS).toEpochMilli());
+
+        resetToDefaults();
+        return transformedValue;
+    }
+
+    boolean hasNotExpired(String key) {
```

Comment:
Not sure how to achieve this, as I explicitly want to be sure the data has not expired. I won't know just with the get, as this is the objective of the lib, to hide the underlying complexity of the cache.

Reply:
I would end up using `java.time.Clock` to control how instant behaves and based on that validate the behavior.

Reply:
and would make sense to test as part of DataStore instead.

Reply:
So basically my suggestion is to test DataStore class directly to validate the functionality of `hasNotExpired` method instead of creating a default method inside BaseProvider which is just used for testing.

Test BaseProvider just for validating fetch from datastore/cache or from remote behavior. And that would simplify test alot.

When it comes to validating without thread.sleep for expiration, if we leverage java.time.clock to manipulate the time, we check the expiration behavior.